File: src/main/java/org/simbrain/util/math/DecayFunction.java
package org.simbrain.util.math;

import org.simbrain.util.UserParameter;
import org.simbrain.util.math.DecayFunctions.GaussianDecayFunction;
import org.simbrain.util.math.DecayFunctions.LinearDecayFunction;
import org.simbrain.util.math.DecayFunctions.QuadraticDecayFunction;
import org.simbrain.util.math.DecayFunctions.StepDecayFunction;
import org.simbrain.util.propertyeditor.CopyableObject;
import org.simbrain.util.stats.ProbabilityDistribution;
import org.simbrain.util.stats.distributions.UniformRealDistribution;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public abstract class DecayFunction implements CopyableObject {

    /**
     * Decay functions for drop-down list used by
     * {@link org.simbrain.util.propertyeditor.ObjectTypeEditor}
     * to set a type of probability distribution.
     */
    public static List<Class> DECAY_FUNCTIONS_LIST = Arrays.asList(
            StepDecayFunction.class,
            LinearDecayFunction.class,
            GaussianDecayFunction.class,
            QuadraticDecayFunction.class
    );

    /**
     * Called via reflection using {@link UserParameter#typeListMethod()}.
     */
    public static List<Class> getTypes() {
        return DECAY_FUNCTIONS_LIST;
    }

    /**
     * If outside of this radius the object has no affect on the network.
     */
    @UserParameter(
            label = "Dispersion",
            description = "If outside of this radius the object has no affect on the network.",
            order = 1)
    private double dispersion = 70;

    /**
     * Peak value.
     */
    @UserParameter(
            label = "Peak Distance",
            description = "Peak value",
            order = 2)
    private double peakDistance = 0;

    /**
     * Noise generator for this decay function if {@link DecayFunction#addNoise} is true.
     */
    @UserParameter(label = "Randomizer", isObjectType = true, order = 1000, tab = "Noise")
    private ProbabilityDistribution randomizer = new UniformRealDistribution();

    /**
     * If true, add noise to object's stimulus vector.
     */
    @UserParameter(
            label = "Add noise",
            description = "If true, add noise to object's stimulus vector.",
            order = 99, tab = "Noise")
    private boolean addNoise = false;

    public double getDispersion() {
        return dispersion;
    }

    public void setDispersion(double dispersion) {
        this.dispersion = dispersion;
    }

    public double getPeakDistance() {
        return peakDistance;
    }

    public void setPeakDistance(double peak) {
        this.peakDistance = peak;
    }

    public ProbabilityDistribution getRandomizer() {
        return randomizer;
    }

    public void setRandomizer(ProbabilityDistribution randomizer) {
        this.randomizer = randomizer;
    }

    public boolean getAddNoise() {
        return addNoise;
    }

    public void setAddNoise(boolean addNoise) {
        this.addNoise = addNoise;
    }

    /**
     * Get the decay amount for the given distance.
     *
     * @param distance the distance to compute the decay scaling factor
     * @return the decay scaling factor
     */
    public abstract double getScalingFactor(double distance);

    /**
     * Apply the scaling factor on a double array based on the given distance.
     *
     * @param distance the distance to compute the decay scaling factor
     * @param vector the vector to be multiplied with the decay scaling factor
     * @return the multiplied vector
     */
    public double[] apply(double distance, double[] vector) {
        double[] ret = new double[vector.length];
        for (int i = 0; i < vector.length; i++) {
            ret[i] = vector[i] * getScalingFactor(distance) + (addNoise ? randomizer.sampleDouble() : 0);
        }
        return ret;
    }

    /**
     * Apply the scaling factor on a Double List based on the given distance.
     *
     * @param distance the distance to compute the decay scaling factor
     * @param vector the vector to be multiplied with the decay scaling factor
     * @return the multiplied vector
     */
    public List<Double> apply(double distance, List<Double> vector) {
        List<Double> ret = new ArrayList<>();
        for (Double e : vector) {
            ret.add(e * getScalingFactor(distance) + (addNoise ? randomizer.sampleDouble() : 0));
        }
        return ret;
    }

    public abstract static class DecayFunctionBuilder<
            B extends DecayFunctionBuilder,
            T extends DecayFunction
            > {

        /**
         * Uniform access to the product being build. Only used in this abstract class
         * where the product cannot be instantiate yet.
         * @return the product being build
         */
        protected abstract T product();

        /**
         * Builds a instance of specific {@link DecayFunction} of given states.
         * @return the final product
         */
        public abstract T build();

        public B dispersion(double dispersion) {
            product().setDispersion(dispersion);
            return (B) this;
        }

        public B peakDistance(double peak) {
            product().setPeakDistance(peak);
            return (B) this;
        }

        public B randomizer(ProbabilityDistribution randomizer) {
            product().setRandomizer(randomizer);
            return (B) this;
        }

        public B addNoise(boolean addNoise) {
            product().setAddNoise(addNoise);
            return (B) this;
        }
    }
}

File: src/main/kotlin/org/simbrain/util/decayfunctions/DecayFunction.kt
package org.simbrain.util.decayfunctions

import org.simbrain.util.UserParameter
import org.simbrain.util.propertyeditor.CopyableObject
import org.simbrain.util.propertyeditor.EditableObject
import kotlin.math.abs

/**
 * Decay functions return a scaling factor between 0 and 1 based on distance.
 *
 * The scaling factor is 1 at the [peakDistance] (default 0) and decreases as distance from the peak increases.
 * Beyond the [dispersion] radius, the scaling factor is 0 (or near 0 for Gaussian).
 *
 * These functions are used to model distance-dependent effects, such as how strongly a spatial object
 * affects a network based on how far away it is. The scaling factor can also be interpreted as a probability.
 *
 * Available implementations:
 * - [StepDecayFunction]: 1 within dispersion, 0 outside
 * - [LinearDecayFunction]: Linear decrease from 1 to 0
 * - [GaussianDecayFunction]: Smooth Gaussian falloff
 */
abstract class DecayFunction(

    /**
     * If outside of this radius the object should have no effect on the network.
     */
    @UserParameter(
        label = "Dispersion",
        description = "If outside of this radius the object has no affect on the network.",
        minimumValue = 0.0,
        order = 1
    )
    var dispersion: Double = 70.0,

    /**
     * The "center" of the decay function where it takes its maximal value.
     */
    @UserParameter(label = "Peak distance", description = "Peak value.", order = 2)
    var peakDistance: Double = 0.0,

    /**
     * Scales the output of the decay function. Can be interpreted as a base probability
     * that is further modulated by distance.
     */
    @UserParameter(
        label = "Base Multiplier",
        description = "Scales the output probability (0-1).",
        minimumValue = 0.0,
        maximumValue = 1.0,
        increment = 0.1,
        order = 3
    )
    var baseMultiplier: Double = 1.0,

) : CopyableObject {

    /**
     * Get the scaling factor for the given distance.
     *
     * Returns a value between 0 and 1 that scales with distance. The scaling factor is 1 at [peakDistance]
     * and decreases as distance increases, reaching 0 at or beyond [dispersion] (or near 0 for Gaussian).
     *
     * Can also be treated as a probability.
     */
    abstract fun getScalingFactor(distance: Double): Double

    // TODO: Stub for future implementation of, for example, elliptical decay functions
    // open fun getScalingFactor(relativeLocation: Point2D): Double {
    //     return 0.0
    // }

    /**
     * Distance from peak.
     *
     * Note that when peak = 0, this is just distance
     */
    fun distanceFromPeak(distance: Double): Double {
        return abs(distance - peakDistance)
    }

    fun copy(copy: DecayFunction): DecayFunction {
        copy.dispersion = dispersion
        copy.peakDistance = peakDistance
        copy.baseMultiplier = baseMultiplier
        return copy
    }

    class DecayFunctionSelector(decayFunction: DecayFunction = LinearDecayFunction(100.0)) : EditableObject {
        /**
         * The layout for the neurons in this group.
         */
        @UserParameter(label = "Decay Function", tab = "Layout", order = 50)
        var decayFunction: DecayFunction = decayFunction

        override val name = "Decay Function"
    }

    override fun getTypeList() = decayFunctionTypes

}

val decayFunctionTypes = listOf(
    StepDecayFunction::class.java,
    LinearDecayFunction::class.java,
    GaussianDecayFunction::class.java,
)